src/lib.rs

```diff
@@ -27,11 +110,197 @@
+            Error::FmtError => write!(f, "formatter error"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for Error {}
+
+/// Common methods needed for formatting _time_.
+pub trait Time {
+    /// Returns the year for _time_ (including the century).
+    fn year(&self) -> i32;
+    /// Returns the month of the year in `1..=12` for _time_.
+    fn month(&self) -> u8;
+    /// Returns the day of the month in `1..=31` for _time_.
+    fn day(&self) -> u8;
+    /// Returns the hour of the day in `0..=23` for _time_.
+    fn hour(&self) -> u8;
+    /// Returns the minute of the hour in `0..=59` for _time_.
+    fn minute(&self) -> u8;
+    /// Returns the second of the minute in `0..=60` for _time_.
+    fn second(&self) -> u8;
+    /// Returns the number of nanoseconds in `0..=999_999_999` for _time_.
+    fn nanoseconds(&self) -> u32;
+    /// Returns an integer representing the day of the week in `0..=6`, with `Sunday == 0`.
+    fn day_of_week(&self) -> u8;
+    /// Returns an integer representing the day of the year in `1..=366`.
+    fn day_of_year(&self) -> u16;
+    /// Returns the number of seconds as a signed integer since the Epoch.
+    fn to_int(&self) -> i64;
+    /// Returns true if the time zone is UTC.
+    fn is_utc(&self) -> bool;
+    /// Returns the offset in seconds between the timezone of _time_ and UTC.
+    fn utc_offset(&self) -> i32;
+    /// Returns the name of the time zone as a string.
+    fn time_zone(&self) -> &str;
+}
+
+/// Provides a buffered `strftime` implementation using a format string with arbitrary bytes.
+pub mod buffered {
+    use super::{Error, Time};
+    use crate::format::TimeFormatter;
+
+    /// Format a _time_ implementation with the specified format byte string,
+    /// writing in the provided buffer and returning the written subslice.
+    ///
+    /// See the [crate-level documentation](crate) for a complete description of possible format specifiers.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use strftime::buffered::strftime;
+    /// use strftime::Time;
+    ///
+    /// // Not shown: create a time implementation with the year 1970
+    /// // let time = ...;
+    /// # include!("mock.rs.in");
+    /// # fn main() -> Result<(), strftime::Error> {
+    /// # let time = MockTime { year: 1970, ..Default::default() };
+    /// assert_eq!(time.year(), 1970);
+    ///
+    /// let mut buf = [0u8; 8];
+    /// assert_eq!(strftime(&time, b"%Y", &mut buf)?, b"1970");
+    /// assert_eq!(buf, *b"1970\0\0\0\0");
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// Can produce an [`Error`](crate::Error) when the formatting fails.
+    pub fn strftime<'a>(
+        time: &impl Time,
+        format: &[u8],
+        buf: &'a mut [u8],
+    ) -> Result<&'a mut [u8], Error> {
+        let len = buf.len();
+
+        let mut cursor = &mut buf[..];
```

Comment:
taking this reference means `Cursor` is a `&mut dyn Write`, correct?

Reply:
Yes, `Write` is only implemented for `&mut [u8]` but not `[u8]`, so the `Write::write` methods take a `&mut &mut [u8]` as its first argument. Only the `cursor` subslice is `dyn Write`: it contains the remaining free part of the `buf` and is shrinked at each call of `Write::write`, eventually returning a `WriteZero` error if the buffer is too small.